Add Screening App (Spanish blueprint)

lfiedlerc opened the pull request:
Add Spanish blueprint which is a screening app that uses mindmeld's slot filling capability.

Commit:
models for screening app

Commit:
Minimum working product for screening app

Commit:
Added implicit yes intents for yes/no questions

Commit:
Refactored screening app to use slot filling.

Commit:
Completed refactoring app to use slot filling and added whatsapp bot server.

Commit:
Removed lint

serapio commented on blueprints/screening_app/__init__.py:
```diff
@@ -0,0 +1,110 @@
+
+    responder.reply(('Bienvenido al sistema de evaluación de salud. Mediante unas'
+                     ' sencillas preguntas, puedo ayudarte a determinar tu riesgo'
+                     ' a padecer prediabetes. Desea conocer su riesgo de padecer prediabetes?'))
```
What is the custom for leaving off the initial '¿'?

lfiedlerc replied:
I believe, grammatically, it should always be used. I simply overlooked it.

serapio commented on blueprints/screening_app/__init__.py:
```diff
@@ -0,0 +1,110 @@
+        elif entity['type'] == 'unit':
+            if entity['role'] == 'height':
+                if entity['value'][0]['cname'] == 'Metros':
+                    if len(str(int(responder.frame[pd.Q_HEIGHT]))) > 1:
```
why compare the string length instead of directly comparing the integer value?

lfiedlerc replied:
Updated to use integer.

serapio commented on blueprints/screening_app/__init__.py:
```diff
@@ -0,0 +1,110 @@
+    if risk >= 5:
+        responder.reply(pd.HIGH_RISK_MSG)
+    else:
+        responder.reply(pd.LOW_RISK_MSG)
```
since you don't use the raw risk score, you could just return a boolean for risk level

lfiedlerc replied:
I updated the function to return a boolean.

ritvikshrivastava commented on blueprints/screening_app/entities/unit/mapping.json:
```diff
@@ -0,0 +1,55 @@
+         "cname":"Kilogramos",
+         "whitelist":[
+            "kilos",
+			   "kilo",
```
I think the indentation is this file got messed up!

lfiedlerc replied:
Fixed.

serapio commented on blueprints/screening_app/domains/prediabetes_screening/answer_age/train.txt:
```diff
@@ -0,0 +1,228 @@
+Tengo {45|sys_number} años de edad
+Yo tengo {45|sys_number} años de edad
+Voy a cumplir {45|sys_number} años de edad
+Cumplo {45|sys_number} años de edadAcabo de cumplir {45|sys_number}
```
mushed 2 together here

lfiedlerc replied:
Fixed.

serapio commented on blueprints/screening_app/domains/prediabetes_screening/answer_age/train.txt:
```diff
@@ -0,0 +1,228 @@
+{45|sys_number}
```
The answer_* intents aren't used anymore, right? We should move the data into opt_in, or remove it if the data is already there.

lfiedlerc replied:
The only reason I kept them was to avoid having a random entity trigger the opt_in intent. For example, if I were to remove answer_age, a random utterance like "32 años" would trigger the form. Right now they go to the default handler.

I did however remove the implicit yes intents since these seem to be more open.

serapio commented on blueprints/screening_app/entities/unit/gazetteer.txt:
```diff
@@ -0,0 +1,25 @@
+metros
+metro
+pies
+pies
```
```suggestion
pie
```

serapio commented on blueprints/screening_app/models/experimentation/intent_model_logreg.pkl.hash:
```diff
@@ -0,0 +1 @@
+6f57c12c2755735e1bb0f828ca884c46b6c7cd81
```
these files should be removed

lfiedlerc replied:
Removed.

ritvikshrivastava commented on blueprints/screening_app/__init__.py:
```diff
@@ -0,0 +1,110 @@
+    If the user accepts the sceening, begin a dialogue flow.
+    """
+
+    for entity in request.entities:
```
How about this?

```
value_map = {'age': pd.Q_AGE, 'weight': pd.Q_WEIGHT, 'height': pd.Q_HEIGHT}
cname_map = {'family_history': pd.Q_FAM, 'hbp': pd.Q_BP, 'active': pd.Q_ACT}

for entity in entities:
    if entity['type'] == 'sys_number':
        responder.frame[value_map[entity['role']]] = entity['value'][0]['value']
    elif entity['type'] == 'unit':
        ...
    elif entity['type'] == 'binary':
        responder.frame[cname_map[entity['role']]] = entity['value'][0]['cname']
    else:
        ....


...
```

ritvikshrivastava replied:
or a common role map for both values and cnames

lfiedlerc replied:
This is much more elegant. Thanks. I have updated the code.

serapio commented on blueprints/screening_app/prediabetes.py:
```diff
@@ -0,0 +1,213 @@
+    '6\'1"': (189, 226),
+    '6\'2"': (194, 232),
+    '6\'3"': (200, 239),
+    '6\'4"': (205, 245)
```
what is the second element of the tuple for? I don't see it used.

lfiedlerc replied:
It's a range. I corrected a mistake in the code that was not using the second value.

serapio commented on blueprints/screening_app/domains/prediabetes_screening/opt_in/train.txt:
```diff
@@ -0,0 +1,1641 @@
+Yo mido {2|sys_number} {metros|unit} y {1|sys_number} {centimetro|unit} de altura
+Yo mido {2|sys_number} {metros|unit} y {1|sys_number} {centimetro|unit} de estatura
+Yo tengo {2|sys_number} {metros|unit} y {1|sys_number} {centimetro|unit} de altura
+Yo tengo {2|sys_number} {metros|unit} y {1|sys_number} {centimetro|unit} de estatura
```
@ritvikshrivastava can the form use this kind of complexity? it would make the experience less robotic.

ritvikshrivastava replied:
Do you mean multiple entities at once? 

If the initial query has multiple entities (like the examples here), they get pre-populated in the form and the user isn't prompted for them.

serapio replied:
This is later in the form, in response to the question "how tall are you?", answering "I am {2|sys_number} {meters|unit} and {1|sys_number} {centimeters|unit} tall". {2 meters} and {1 centimeter} are known distance entities in English, but not in Spanish, and the combined phrase is not recognized in either language.

ritvikshrivastava replied:
Hmm, I'll look if that initial fill function can be used in between here. 

Alternatively, one direct way currently could be to prompt the message only asking the weight/height in a specific unit. say Kilos/Metres? The app logic can then be calculated using that selected unit always as the onus on replying to the prompt with that unit is on the user.

Another way could be to have a slot for the measurement system they want to use, and then accordingly give a followup form to fill values in kilos/metres or pounds/feet.

ritvikshrivastava replied:
Can be resolved using https://github.com/lfiedlerc/mindmeld-blueprints/pull/1 and https://github.com/cisco/mindmeld/pull/205.

serapio commented on blueprints/screening_app/domains/prediabetes_screening/answer_yes_active/train.txt:
```diff
@@ -0,0 +1,102 @@
+hago mucho deporte
```
@ritvikshrivastava is there a way to incorporate this kind of flexibility into the slot filling? the answer_yes_* and answer_no_* intents are pragmatically natural responses to the yes/no questions, but don't explicitly say yes or no. adding these phrases into an entity map would work, it seems, but would be more brittle.

ritvikshrivastava replied:
@serapio They aren't being used now, right? Slot-filling essentially implies getting rid of the binary part and asking for the value outright?

ritvikshrivastava replied:
One way would be to define a custom validation function and checking against these lists of queries for each of the corresponding slots

serapio reviewed (changes requested):
Should drop the pickle files from the branch. Maybe we should add *.pkl to gitignore.

Commit:
Updated .gitignore to include .pkl* files

Corrected mapping formatting for unit entity

Added newline between lumped queries

Code cleanup

Removed implicit yes intents

lfiedlerc committed:
Update blueprints/screening_app/entities/unit/gazetteer.txt

Co-authored-by: Lucien Carroll <lucien@discurs.us>

Commit:
Merge branch 'Screening-with-slot-filling' of github.com:lfiedlerc/mindmeld-blueprints into Screening-with-slot-filling

serapio commented on blueprints/screening_app/__init__.py:
```diff
@@ -0,0 +1,101 @@
+
+    # Respond with a random selection from one of the canned "goodbye" responses.
+    responder.reply(['Gracias por su visita. Adios!',
+                     'Gracias por su visita. Hasta luego!',
```
¡ here

lfiedlerc replied:
Added.

Commit:
Added '¡' to goodbye responses.

ritvikshrivastava committed:
using custom resolution/validation for height/weight multi-entity queries

ritvikshrivastava committed:
remove print statement

ritvikshrivastava commented:
Hey @lfiedlerc, these changes https://github.com/lfiedlerc/mindmeld-blueprints/pull/1 can be added to this blueprint now. They incorporate @serapio's multi-entity handling suggestion for a slot!

lfiedlerc committed:
Merge pull request #1 from ritvikshrivastava/spanish-blueprint-restructuring

Spanish blueprint restructuring

Commit:
Removed lint

lfiedlerc commented:
After discussing with @serapio, we determined that slot filling was not an optimal solution for this app. We will revert to using a dialogue flow.
